library/src/main/java/com/novoda/downloadmanager/lib/jobscheduler/DownloadJob.java

```diff
@@ -0,0 +1,121 @@
+package com.novoda.downloadmanager.lib.jobscheduler;
+
+import android.os.Handler;
+import android.os.Looper;
+import android.support.annotation.NonNull;
+
+import com.evernote.android.job.Job;
+import com.evernote.android.job.JobRequest;
+import com.evernote.android.job.util.support.PersistableBundleCompat;
+import com.novoda.downloadmanager.lib.DownloadServiceJob;
+import com.novoda.downloadmanager.lib.DownloadStatus;
+import com.novoda.downloadmanager.lib.logger.LLog;
+
+import java.util.concurrent.TimeUnit;
+
+public class DownloadJob extends Job {
+
+    public static final String TAG = "download_job_tag";
+
+    private static final String ENFORCE_EXECUTION = "enforceExecution";
+    private static final boolean STANDARD_EXECUTION = false;
+    private static final boolean ENFORCED_EXECUTION = true;
+    private static final long BACKOFF_MILLIS = TimeUnit.SECONDS.toMillis(5);
+    private static final long EXECUTION_START_MILLIS = TimeUnit.SECONDS.toMillis(1);
+    private static final long EXECUTION_END_MILLIS = TimeUnit.SECONDS.toMillis(2);
+    private static final Object LOCK = new Object();
+
+    @NonNull
+    @Override
+    protected Result onRunJob(Params params) {
+        LLog.v("job starts right now");
+
+        final DownloadServiceJob[] downloadServiceJob = new DownloadServiceJob[1];
+
+        ensureDownloadServiceJobInstanceExists(downloadServiceJob);
+        waitForDownloadServiceJobInstanceToBeReady();
+
+        PersistableBundleCompat extras = params.getExtras();
+        boolean enforceExecution = extras.getBoolean(ENFORCE_EXECUTION, STANDARD_EXECUTION);
+
+        int status = downloadServiceJob[0].onStartCommand(enforceExecution);
+
+        if (jobHasSucceeded(status)) {
+            LLog.v("job is completed");
+            return Result.SUCCESS;
+        } else if (jobNeedsRescheduling(status)) {
+            LLog.v("job is going to be rescheduled");
+            return Result.RESCHEDULE;
+        } else {
+            LLog.v("job failure");
+            return Result.FAILURE;
+        }
+    }
+
+    private void ensureDownloadServiceJobInstanceExists(final DownloadServiceJob[] downloadServiceJob) {
+        new Handler(Looper.getMainLooper()).post(new Runnable() {
+            @Override
+            public void run() {
+                downloadServiceJob[0] = DownloadServiceJob.getInstance();
```

Comment:
`.getInstance()` needs to happen in the main thread due to content provider instantiations